### PR TITLE
Add training performance investigation script (ONNX graph analyzer)

### PR DESCRIPTION
**Pull request by kit1980**
Here is an example output for a model I tried:

```
$ python performance_investigation.py o-_optimized_training.onnx 
Processing o-_optimized_training.onnx...
ATenOp(s) found:
----------
ATenOp_39: aten::embedding
ATenOp_39_Grad/ATenOp_3: aten::embedding_backward
ATenOp_31_Grad/ATenOp_3: aten::embedding_backward
ATenOp_35_Grad/ATenOp_3: aten::embedding_backward
ATenOp_31: aten::embedding
ATenOp_35: aten::embedding
----------
Examine whether Dropout_137 should be fused with the leading Add_134 op into BiasDropout node.
Examine whether Dropout_159 should be fused with the leading Add_156 op into BiasDropout node.
Examine whether Dropout_253 should be fused with the leading Add_250 op into BiasDropout node.
Examine whether Dropout_275 should be fused with the leading Add_272 op into BiasDropout node.
Examine whether Softmax_230 should be fused with the leading Add_229 op into BiasSoftmax node.
Examine whether Softmax_114 should be fused with the leading Add_113 op into BiasSoftmax node.
```

### CODE REVIEW

File: orttraining/tools/scripts/performance_investigation.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    aten_ops = []
+    msgs = []
+
+    visited = set([])
+    def dfs(node, path):
+        if node.name in visited:
+            return
+        visited.add(node.name)
+
+        if node.op_type == "ATenOp":
```
Comment (SherlockNoMad) [inline orttraining/tools/scripts/performance_investigation.py:33]: Also need for PythonOp.

File: orttraining/tools/scripts/performance_investigation.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        for node in input_to_nodes.get(i.name, []):
+            if node.name not in visited:
+                dfs(node, [])
+
+    for i in model.graph.input:
+        if i.name in output_to_node:
+            continue
+        for node in input_to_nodes.get(i.name, []):
+            if node.name not in visited:
+                dfs(node, [])
```
Comment (SherlockNoMad) [inline orttraining/tools/scripts/performance_investigation.py:98]: As there are many connection edges between forwards nodes and backward nodes in a training graph, with dfs traversal, it's likely that you jump directly from forward node to a backward node, and miss the pattern you are looking for. 

Actually, there is no single graph transversal algo that give you a reliable serialization that you can match with a linear pattern.

**Comment by kit1980** [comment]
Closing this PR and will create a new one per offline conversation with Sherlock.

File: orttraining/tools/scripts/performance_investigation.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+def process_file(onnx_file, torch_exported_file=False, pre_grad_optimized_file=False, optimized_file=False, execution_model_file=False):
+    print(f"Processing {onnx_file}...")
+    model = onnx.load(onnx_file)
+
+    input_to_nodes = {}
+    output_to_node = {}
+    for node in model.graph.node:
+        for i in node.input:
+            input_to_nodes[i] = input_to_nodes.get(i, []) + [node]
+        for o in node.output:
```
[the remark this answers is not in the document]
Comment (kit1980) [reply orttraining/tools/scripts/performance_investigation.py:21]: That's what already happening in the code, no?
`input_to_nodes[i]` (consumers) is map <NodeArg, vector>, and `output_to_node[o]` (producers) is  map<NodeArg, Node>.

**Comment by kit1980** [comment]
New PR: https://github.com/microsoft/onnxruntime/pull/9791

**Pull request closed without merging**